### PR TITLE
Remove pug 'index' template and replace with JSX

### PR DESCRIPTION
**Pull request by flootr**
_previous try: https://github.com/Automattic/wp-calypso/pull/21365_

Follow-up of #8459, this PR basically is intended to continue the work @ockham already did in his PR. For now it **only replaces the 'index' template with JSX**. Let's see how this goes and iterate on that. Then replace the other templates stepwise.

Also a stepping-stone towards #19833.

### todos

- [x] Render the `index` template via React instead of Pug
- [x] Cover all changes by tests
- [x] Add a `<Head />` component to unify head sections

### caveats

- We are calling `renderToString()` for SSRed content and additionally `renderToStaticMarkup()` for the template. Should probably only be one call. _Please refer to https://github.com/Automattic/wp-calypso/pull/8459#issue-181002246 for why this wasn't done in the first place._

### CODE REVIEW

**Comment by dmsnell** [comment]
I tried following the chain of related PRs (thanks for providing those!) but didn't easily find an answer to what I think is the most important question here: why are we doing this? what do we expect to gain from it? what drawbacks do we think it will create?

**Comment by flootr** [comment]
Thanks @dmsnell for pointing that out. I totally agree with you that there should be an actual problem which got solved. Here's what I think about that:

- We are currently using 2 different rendering engines (Pug and React) to render on the server (Pug for the shells and React for SSR). This is not ideal imo as it's confusing and creates complexity because there's a few steps more needed in order to do so. Switching to React should enable us to combine those steps (as needed) in the long term (see the _caveat_ I mentioned above).
- In order to improve the app shell (#19833) it makes sense to me to use components which are already used on the client side (and for SSR). Like the [`<Masterbar />`](https://github.com/Automattic/wp-calypso/pull/21424/files#diff-3fb371c23faa311830bf33b7e7e855a3R94). This reduces redundancy as we don't re-build stuff in Pug which are already there (written in React). 

I tbh don't know how React performs against Pug in (static) rendering. But I'm eager to find out if that's something which is of any concern. Also React enables us to do things like [streaming](https://github.com/Automattic/wp-calypso/issues/19822#issuecomment-348177437) which is currently not possible with Pug.

I hope that gets you a bit more insight in why I am doing this refactoring. If there's still something which is not clear or I am missing I'm happy to elaborate more.

**Comment by flootr** [comment]
@dmsnell I've tested a bit and it seems that Pug takes (on my local machine) at least ~60ms to render the markup and React (with `renderToStaticMarkup()`) about ~2ms.

Here's what I did to measure:

```js
// server/render/index.js
export function renderJsx( view, props ) {
        // ...
	console.time();
	const markup =
		doctype + ReactDomServer.renderToStaticMarkup( React.createElement( component, props ) );
	console.timeEnd();
	return markup;
}

export function serverRender( req, res ) {
	// ...
	if ( req.query.engine === 'pug' ) {
		console.time();
		res.render( 'index', context, ( err, html ) => {
			console.timeEnd();
			res.send( html );
		} );
	} else {
		res.send( renderJsx( 'index', context ) );
	}
```

The query `?engine=pug` determines whether the template should be rendered with Pug instead of JSX. The snippets from above is meant to be added to the code from this PR.

Here are the times (in sequential order)

```
// Pug (Express integration)
default: 328.218ms
default: 154.045ms
default: 138.716ms
default: 120.669ms
default: 138.593ms
default: 109.614ms
default: 99.813ms
default: 96.259ms
default: 98.515ms
default: 72.774ms
default: 74.310ms
default: 75.304ms
default: 61.625ms
default: 59.123ms
default: 54.335ms
default: 72.487ms
default: 69.025ms
default: 72.568ms
default: 82.825ms
default: 52.947ms
default: 51.739ms
default: 61.100ms
default: 64.608ms

// React with renderToStaticMarkup()
default: 6.831ms
default: 1.898ms
default: 1.527ms
default: 1.667ms
default: 1.496ms
default: 5.509ms
default: 1.526ms
default: 2.659ms
default: 1.961ms
default: 2.076ms
default: 1.916ms
default: 2.046ms
default: 2.348ms
default: 1.874ms
default: 2.080ms
default: 1.818ms
default: 1.398ms
default: 5.633ms
default: 1.242ms
default: 1.860ms
default: 6.420ms
default: 1.641ms
```

It would be interesting if anyone could confirm those numbers.

**Comment by dmsnell** [comment]
those are great numbers @flootr! since we're just measuring the time of render does it take into account the initialization times of `pug` vs. `react`? is React booting up _anyway_ on SSR?

**Comment by flootr** [comment]
> since we're just measuring the time of render does it take into account the initialization times of pug vs. react?

Not sure what you mean by initialization times. I assume it's the `require()`?

The following script:
```js
// bootup.js
console.time('pug')
const pug = require('pug')
console.timeEnd('pug')
console.time('react')
const react = require('react')
console.timeEnd('react')
```

gave me the folllowing results (several tests were ~ the same):

```bash
$ node bootup.js
pug: 118.472ms
react: 2.931ms
```

Not sure if there are any other initializations going on for those libs tbh 🤔

**Comment by dmsnell** [comment]
@flootr you are so awesome - thanks for posting those tests. if these tests are reliable, and I don't have much reason to doubt that they are, then this PR could make a substantial improvement on our SSR stuff.

@ockham did you get a chance to look at this yet? I think it would be great to double-verify that this is all green-lighted, that we won't be accidentally harming ourselves by removing `pug`. it _seems_ clear that having _two_ systems is clearly worse than _one_, but I don't know my way around here like you do.

**Comment by flootr** [comment]
Thank you @dmsnell. Regarding my last comment I want to add that even if React is significantly loaded faster with `require` this does not have much impact on the runtime I guess, because this likely happens only once (when starting the node process) and `require` uses caching. So loaded once the times are ~ the same.

```
$ node bootup.js
react-1: 4.545ms
react-2: 0.019ms
pug-1: 125.687ms
pug-2: 0.013ms
```

**Comment by ockham** [comment]
I'll try to give this a closer look this week. Apologies for the delay!

**Comment by dmsnell** [comment]
Is there a need to split out the `<Head>` component? I don't have strong feelings about it but in some regards having the `index` as one big file makes sense to me so we can see everything that's happening. (this would imply having one big `render()` too and not calling `{ this.renderHead() }` as well. Thoughts?

**Comment by flootr** [comment]
@dmsnell it makes sense to me to have a separate component which can, eventually, be used in other templates as well (thinking of `404`, `500`, `desktop`, `browsehappy`) so these don't get out of sync.

File: client/components/head/index.jsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+			<link
+				rel="shortcut icon"
+				type="image/vnd.microsoft.icon"
+				href={ faviconURL }
+				sizes="16x16 32x32"
+			/>
+			<link rel="shortcut icon" type="image/x-icon" href={ faviconURL } sizes="16x16 32x32" />
+			<link rel="icon" type="image/x-icon" href={ faviconURL } sizes="16x16 32x32" />
+
+			<link
```
Comment (samouri) [inline client/components/head/index.jsx]: We could cut out ~100 lines with something like:
```jsx
[64, 96, 192, ...].map( px => (
  <link 
     rel="icon" 
     type="image/png"
     href={cdn + `/i/favicons/apple-touch-icon-${px}x${px}.png`}
))
```

Comment (flootr) [reply client/components/head/index.jsx]: True, the first ones have a slightly different path but still that would save some lines. As this PR is already merged I'm going to address this in a follow-up PR if that's okay?

**Comment by ockham** [comment]
Awesome work @flootr, thanks a lot, and sorry for not having been able to review this one!

**Pull request merged**